L58-python-async-api.md

```diff
@@ -0,0 +1,453 @@
+> This way alternative event loop implementations can inject their own optimized
+> implementations of a Future object.
+
+### Python Coroutines in `asyncio`
+
+In a single-threaded application, creating a coroutine object doesn't
+necessarily mean it is scheduled to be executed in the event loop.
+
+The functions defined by `async def`, underlying, is a function that returns a
+Python generator. If the program calls an `async def` function, it will NOT be
+executed. This behavior is one of the main reason why mixing `async def`
+function with normal function is a bad idea.
+
+There are three mechanisms to schedule coroutines:
+
+1. Await the coroutine `await asyncio.sleep(1)`;
+2. Submit the coroutine to the event loop object `loop.call_soon(coro)`;
+3. Turn coroutine into an `asyncio.Task` object `asyncio.ensure_future(coro)`.
+
+## Proposal
+
+This gRFC intended to fully utilize `asyncio` and create a Pythonic paradigm for
+gRPC Python programming in Python 3 runtime.
+
+### Two Sets of API In One Package
+
+The new API will be isolated from the current API. The implementation of the new
+API is an entirely different stack than the current stack. On the downside, most
+gRPC objects and definitions can't be shared between these two sets of API.
+After all, it is not a good practice to introduce uncertainty (return coroutine
+or regular object) to our API. This decision is made to respect our contract of
+API since GA in 2016, including the underlying behaviors. **Developers who use
+our current API should not be affected by this effort.**
+
+For users who want to migrate to new API, the granularity migration is per
+channel, per server level. For both channel-side and server-side, the adoption
+of `asyncio` does not only include cosmetic changes of adding `async`/`await`
+keywords throughout the code base, it also requires thinking about the change
+from a potentially multi-threaded application to a single-threaded asynchronous
+application. For example, the synchronization methods have to be changed to use
+the `asyncio` edition (see [Synchronization
+Primitives](https://docs.python.org/3/library/asyncio-sync.html)).
+
+### Overview of Async API
+
+All `asyncio` related API will be kept inside the `grpc.aio` module. Take the
+most common use case as an example, the new API will diverge from the old API
+since the creation of `Channel` and `Server` object. New creation methods are
+added to instantiate async `Server` or `Channel` object:
+
+```Python
+import grpc
+server = grpc.aio.server(...)
+channel = grpc.aio.insecure_channel(...)
+channel = grpc.aio.secure_channel(...)
+channel = grpc.aio.intercept_channel(...)
+```
+
+To reduce cognitive burden, the new `asyncio` API should share the same
+parameters as the current API, and keep the usage similar, except the following
+cases:
+
+* `grpc.aio.server` no longer require the application to provide a
+  `ThreadPoolExecutor`;
+* Interfaces returning `Future` object are replaced.
```

Comment:
or removed since `asyncio` already provides a way for running a coroutine as an independent task, and implementing the `Future` contract, for example:

```python
task = asyncio.run(stub.Hi(echo_pb2.EchoRequest(message="ping")))
task.add_done_callback(my_callback)
```

Reply:
Yes, so for those use case, they should return a Task or our context class instead. The idea being, if we must use `Future` mechanism, we should hide the implementation details. Otherwise, we should yield the choice of `Future` implementation to our users.

https://docs.python.org/3.7/library/asyncio-future.html#asyncio.Future

Reply:
s/asyncio.run/asyncio.ensure_future

Reply:
So, either the user uses the Asyncio API directly by scheduling a separated task or the `aio` library does it internally and returns either an `Asyncio.Task` [1] or the context class.

Seems that I will go for the second option only if the object that needs to be returned is not the `Asyncio.Task` one. Otherwise we will be adding a superfluous wrapper IMO.

If we return a gRPC object this would need to implement all of the interfaces - like being awaitable - to be used by the set of the asyncio methods that deal with awaitable objects, for example `wait` [2] and `wait_for` [3]. Otherwise, the user would have a lot of friction using these asyncio methods when a gRPC future object is given as a parameter, for example:

```python
grpc_task = stub.Hi.future(echo_pb2.EchoRequest(message="ping"))
await asyncio.wait([grpc_task])
```

This makes me think that the simpler the better, so going for the solution where the `aio` module does not implement any ad-hoc method, and the user uses the Asyncio API directly. Going for the ad-hoc alternative would need a strong reason. 

[1] https://docs.python.org/3/library/asyncio-task.html#asyncio.Task
[2] https://docs.python.org/3/library/asyncio-task.html#asyncio.wait
[3] https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for

Reply:
I'm not sure I fully understand your argument. What is your conclusion here? Do you support returning an `asyncio.Task` object?

Reply:
second thought, and related with this comment [1].

Having the feeling that we would need to return a custom object where internalities will be hidden. For example for the use case of the `future()` method of the unary call, we would return a `grpc.aio.Call` object which would implement the `__await__` method, making it compatible with Asyncio API primitives that are expecting _awaitables_ like `wait`, `wait_for`, `ensure_future` or the `await` keyword, so the following should work - I haven't tried it:

```python
call = stub.Hi.future(...)
await call
```
Ofc the following snippet will have the same effect, since the object returned implement the __await__ method

```python
call = await stub.Hi.future(...)
```


Internally the `grpc.aio.Call` wold start the task in the background, having the chance of using the methods provided by the `asyncio.Task` for implementing some of the public methods of the `grpc.aio.Call` object, for example

```python
class Call:
    def __init__(coro,......):
        self._task = asyncio.create_task(coro)
        ....
    def __await__(self):
        return self._task.__await__()
   async def cancel(self):
        self._task.cancel()
        ....
```

[1] https://github.com/grpc/proposal/pull/155/files#r311288205

Reply:
How about make the `grpc.aio.Call` class inherit from `asyncio.Task`? So we get all the subtle changes across multiple Python versions, and also get the chance to inject the logic we want?

```Python
class Call(asyncio.Task):
    def __init__(coro, ...):
        ....
```

Reply:
Inheriting directly from the `asyncio.Task` would put our API at merci of the Asyncio maintainers. If a new method is added for the `asyncio.Task` object, will we have 100% confidence that is fully supported in our use case? I don't think so. While not exposing the task gives us the chance and freedom of using the methods that we want and in the way that we want.

But this discussion might be an implementation detail.

Reply:
I think both direction is fine for me. My original intent to propose inherit from `Task` is to reduce our maintenance burden. It can cut down the boilerplates we have write in `grpc.aio.Call` class. The person who implements it can decide which way to go.

Also, there is another upside that users can use `isinstance(call, asyncio.Task)` logic.

Reply:
I'm in favor of resolving this during the implementation :)